### DISCUSSION TITLE
How to setup xcode related env in aspect rule?

### DISCUSSION THREAD

**Discussion by kkpattern** [bazelbuild/rules_apple#1566] [Q&A] [answered]
Hi, I'm trying to use [bazel_clang_tidy](https://github.com/erenon/bazel_clang_tidy) to run clang-tidy on my `objc_library` targets. Clang-tidy failed to find the system headers. I think this is because bazel use `__BAZEL_XCODE_SDKROOT__` in compile flags then replace it later with `SDKROOT` in `wrapped_cc`. However, when I build the aspect rule, the `SDKROOT` env is not set. Is there a way I can setup XCode related env in aspect rule?

Thanks!

**Comment by keith** [comment, marked as answer]
I haven't done this with aspects before but you should checkout this pattern https://github.com/bazelbuild/rules_apple/blob/89b1094e58433dd742b630da0fb95a67979f682a/apple/testing/default_runner/watchos_test_runner.bzl#L108-L113 and how that is used. You might have to do the replacements yourself depending on the case.

**Reply by kkpattern** [reply]
> I haven't done this with aspects before but you should checkout this pattern
> 
> https://github.com/bazelbuild/rules_apple/blob/89b1094e58433dd742b630da0fb95a67979f682a/apple/testing/default_runner/watchos_test_runner.bzl#L108-L113
> 
> and how that is used. You might have to do the replacements yourself depending on the case.

Thanks! It worked!

So when execute actions bazel will [put `SDKROOT` into env]( https://github.com/bazelbuild/bazel/blob/ba1ef9261da4d3c3e93ac1bedc85d17016ef4ac4/src/main/java/com/google/devtools/build/lib/exec/local/XcodeLocalEnvProvider.java#L108). But it requires xcode config related env variables are set first. So we need first to add `_xcode_config` attr and `apple` fragment to the aspect rule:

```python
clang_tidy_aspect = aspect(
    implementation = _clang_tidy_aspect_impl,
    fragments = ["cpp", "apple", "objc"],
    attrs = {
        ...
        "_xcode_config": attr.label(
            default = configuration_field(
                fragment = "apple",
                name = "xcode_config_label",
            ),
        ),
    },
)
```

Then we get the required env variables by:

```python
    env = {}
    xcode_config = ctx.attr._xcode_config[apple_common.XcodeVersionConfig]
    env.update(apple_common.apple_host_system_env(xcode_config))
    env.update(apple_common.target_apple_env(xcode_config, ctx.fragments.apple.single_arch_platform))
```

Finally, we put the env into `action.run`:

```python
    ctx.actions.run(
        ...
        env = env,
    )
```

Now we have `SDKROOT` and `DEVELOPER_DIR` setup for actions.

I also found [`apple_support`](https://github.com/bazelbuild/apple_support/blob/master/doc/apple_support.md#apple_support.run) already took care of this for us. I encounter a problem when trying `apple_support` so end up doing it on my own. But if anyone need to run actions with `SDKROOT` and `DEVELOPER_DIR` can give it a try.

@keith thanks for your help!
